src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala

```diff
@@ -141,81 +148,81 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
               case Success(None) =>
                 // No remote found, return local, async update both
                 reporter.onCacheMissButFoundExpiredLocal(key, elapsedTime(startTime))
                 tryGenerateAndSet(key, genValue, startTime)
-                Future.successful(expiredLocalValue.value)
+                Future.successful(expiredLocalValue.value.get)
               case Failure(e) =>
                 reporter.onRemoteError(key, e, elapsedTime(startTime))
                 logger.warn(s"apply, key: $key expired local value and failed to get remote", e)
                 tryGenerateAndSet(key, genValue, startTime)
-                Future.successful(expiredLocalValue.value)
+                Future.successful(expiredLocalValue.value.get)
             }
           case Success(expiredLocalValue) if remoteRW.isEmpty =>
             // There is no remote cache configured, we'are on our own
             // Return expired value and try to generate a new one for the future
             reporter.onCacheMissButFoundExpiredLocal(key, elapsedTime(startTime))
             tryGenerateAndSet(key, genValue, startTime)
-            Future.successful(expiredLocalValue.value)
+            Future.successful(expiredLocalValue.value.get)
           case Failure(e) =>
             // This is almost impossible to happen because it's local and we don't save failed values
             reporter.onLocalError(key, e, elapsedTime(startTime))
             logger.warn(s"apply, key: $key got a failed future from cache!? This is almost impossible!", e)
-            tryGenerateAndSet(key, genValue, startTime).map(_.value)
+            tryGenerateAndSet(key, genValue, startTime).map(_.value.get)
         }
       case None if remoteRW.nonEmpty =>
         // No local, let's try remote
         remoteRW.get.get(key).asTry().flatMap {
-          case Success(Some(remoteValue)) if !remoteValue.hasExpired(ttl, now) =>
+          case Success(Some(remoteValue)) if !remoteValue.hasExpired(ttl, now, ttlCachedErrors) =>
             // Remote is good, set locally and return it
             reporter.onRemoteCacheHit(key, elapsedTime(startTime))
             localCache.foreach(_.set(key, remoteValue))
-            Future.successful(remoteValue.value)
+            Future.successful(remoteValue.value.get)
           case Success(Some(expiredRemote)) =>
             // Expired remote, return the it, async update
             reporter.onCacheMissButFoundExpiredRemote(key, elapsedTime(startTime))
             tryGenerateAndSet(key, genValue, startTime).map(_.value)
-            Future.successful(expiredRemote.value)
+            Future.successful(expiredRemote.value.get)
           case Success(None) =>
             // No good remote, sync generate
             reporter.onCacheMissNothingFound(key, elapsedTime(startTime))
-            tryGenerateAndSet(key, genValue, startTime).map(_.value)
+            tryGenerateAndSet(key, genValue, startTime).map(_.value.get)
```

Comment:
@fernandors87 nice catch about `Future.fromTry`.
And about this point in the code?

What do you think about this solution?

```
tryGenerateAndSet(key, genValue, startTime).flatMap(tsv => Future.fromTry(tsv.value))
```

I am not sure about this `flatMap`. Is it blocking?

Reply:
tryGenerateAndSet returns a Future.
All future ops are non blocking.

Another way to write the same code:

```scala
for {
  tsv <- tryGenerateAndSet(key, genValue, startTime)
  value <- Future.fromTry(tsv.value)
} yield value